src/OpenTelemetry.Instrumentation.AspNetCore/TracerProviderBuilderExtensions.cs

```diff
@@ -51,22 +53,45 @@ public static class TracerProviderBuilderExtensions
 
         internal static TracerProviderBuilder AddAspNetCoreInstrumentation(
             this TracerProviderBuilder builder,
-            AspNetCoreInstrumentation instrumentation)
+            AspNetCoreInstrumentation instrumentation,
+            IServiceProvider serviceProvider = null)
         {
-            builder.AddSource(HttpInListener.ActivitySourceName);
-            builder.AddLegacySource(HttpInListener.ActivityOperationName); // for the activities created by AspNetCore
+            // For .NET7.0 onwards activity will be created using activitySource.
+            // https://github.com/dotnet/aspnetcore/blob/bf3352f2422bf16fa3ca49021f0e31961ce525eb/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L327
+            // For .NET6.0 and below, we will continue to use legacy way.
+            if (HttpInListener.IsNet7OrGreater)
+            {
+                var activitySourceService = serviceProvider?.GetService<ActivitySource>();
```

Comment:
Not sure if this is safe. Anyone can add `ActivitySource` to DI, and how do we know if this is the one added by asp.net core?

Reply:
This is how library passes the activitysource as well to diagnostics module. https://github.com/dotnet/aspnetcore/blob/0e5cbf447d329a1e7d69932c3decd1c70a00fbba/src/Hosting/Hosting/src/Internal/WebHost.cs#L152